### PR TITLE
Fix panic in when a service tailed by the agent is deleted

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

This panic could be triggered when deleting a pod that was currently being tailed by the log agent.

By using a slice with a same underlying array we can remove item without triggering "panic slice bounds out of range".

### Describe how to test/QA your changes

Enable collecting logs from all containers in the agent, and check that the stopping a container service is capture by the agent.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/logs/service/services.go [hunk 1]
```diff
@@ -49,11 +49,13 @@ func (s *Services) RemoveService(service *Service) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	for i, svc := range s.services {
-		if svc.Type == service.Type && svc.Identifier == service.Identifier {
-			s.services = append(s.services[:i], s.services[i+1:]...)
+	remainingServices := s.services[:0]
+	for _, svc := range s.services {
+		if svc.Type != service.Type && svc.Identifier != service.Identifier {
```
Comment (vickenty) [inline pkg/logs/service/services.go:54]: Wouldn't this also remove all other services with the same type as the one being removed?

**Review by vickenty — COMMENTED** [review]
Would it be possible to add tests for this?

**Review by maycmlee — APPROVED** [review]
👍 for docs.

**Review by djmitche — APPROVED** [review]

**Review by vickenty — CHANGES_REQUESTED** [review]
`s.services = append(s.services[:i], s.services[i+1:]...)` is fine even if `i` points to the last element: https://go.dev/play/p/J3KTR6NkI-d. The problem is either elsewhere, or this is a 
data race: in either case, the proposed change won't fix it.

**Comment by hush-hush** [comment]
> `s.services = append(s.services[:i], s.services[i+1:]...)` is fine even if `i` points to the last element: https://go.dev/play/p/J3KTR6NkI-d. The problem is either elsewhere, or this is a data race: in either case, the proposed change won't fix it.

The issue is when you delete more than 1 entry from the list: https://go.dev/play/p/LAcL6OvHBTd

Which it what happened from the panic we received:
```
panic: runtime error: slice bounds out of range [15:14]

goroutine 481 [running]:

github.com/DataDog/datadog-agent/pkg/logs/service.(*Services).RemoveService(0xc0010b49c0, 0xc000b9d4c0)

        /omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/service/services.go:50 +0x425

github.com/DataDog/datadog-agent/pkg/logs/schedulers.(*sourceManager).RemoveService(0x0, 0x0)

        /omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/schedulers/source_manager.go:46 +0x1d

github.com/DataDog/datadog-agent/pkg/logs/schedulers/ad.(*Scheduler).Unschedule(0xc000b8f860, {0xc0011d3b00, 0x1, 0x0})

        /omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/schedulers/ad/scheduler.go:149 +0x307

github.com/DataDog/datadog-agent/pkg/logs/internal/util/adlistener.(*ADListener).Unschedule(0xc000f79b10, {0xc0011d3b00, 0x1, 0x7fec84ef62b8})

        /omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/internal/util/adlistener/ad.go:66 +0x2b

github.com/DataDog/datadog-agent/pkg/autodiscovery/scheduler.(*MetaScheduler).Unschedule(0xc0001b1320, {0xc0011d3b00, 0xc0011d3b00, 0x1})

        /omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/scheduler/meta.go:88 +0x12e

github.com/DataDog/datadog-agent/pkg/autodiscovery.(*AutoConfig).applyChanges(0xc000b91d60, {{0x0, 0x0, 0x0}, {0xc0011d3b00, 0x1, 0x1}})

        /omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:501 +0x59

github.com/DataDog/datadog-agent/pkg/autodiscovery.(*AutoConfig).processDelService(0xc000b91d60, {0x513b2c0, 0xc00176d050})

        /omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:480 +0x305

github.com/DataDog/datadog-agent/pkg/autodiscovery.(*AutoConfig).serviceListening(0xc000b91d60)

        /omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:116 +0x290

created by github.com/DataDog/datadog-agent/pkg/autodiscovery.NewAutoConfig

        /omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:66 +0x65
```

**Comment by ykyr** [comment]
I didn't find the related Issue and since the PR is prepared, just gonna add here.
We deployed DD agent `datadoghq/agent:7.39.1`  and on all the clusters `agent` container is restarting with the failing readiness probe 
```
kubelet  Readiness probe failed: HTTP probe failed with statuscode: 500
```

The panic that agent throws before the shutdown is exactly the one commented above. 
```
panic: runtime error: slice bounds out of range [69:68]

goroutine 394 [running]:
github.com/DataDog/datadog-agent/pkg/logs/service.(*Services).RemoveService(0xc0015bbf20, 0xc0007d9160)
	/omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/service/services.go:54 +0x475
github.com/DataDog/datadog-agent/pkg/logs/schedulers.(*sourceManager).RemoveService(0x0, 0x0)
	/omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/schedulers/source_manager.go:46 +0x1d
github.com/DataDog/datadog-agent/pkg/logs/schedulers/ad.(*Scheduler).Unschedule(0xc0014aaa38, {0xc00251f600, 0x1, 0x0})
	/omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/schedulers/ad/scheduler.go:151 +0x4e7
github.com/DataDog/datadog-agent/pkg/logs/internal/util/adlistener.(*ADListener).Unschedule(0xc000f9f6c8, {0xc00251f600, 0x1, 0x1})
	/omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/internal/util/adlistener/ad.go:66 +0x2b
github.com/DataDog/datadog-agent/pkg/autodiscovery/scheduler.(*MetaScheduler).Unschedule(0xc001626a08, {0xc00251f600, 0x1947af4, 0x1})
	/omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/scheduler/meta.go:94 +0x12e
github.com/DataDog/datadog-agent/pkg/autodiscovery.(*AutoConfig).applyChanges(0xc000d9e780, {{0x0, 0x0, 0x0}, {0xc00251f600, 0x1, 0x1}})
	/omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:479 +0x3e6
github.com/DataDog/datadog-agent/pkg/autodiscovery.(*AutoConfig).processDelService(0xc000d9e780, {0x555e7b8, 0xc0017e7140}, {0x55cff40, 0xc0028121b0})
	/omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:454 +0x346
github.com/DataDog/datadog-agent/pkg/autodiscovery.(*AutoConfig).serviceListening(0xc000d9e780)
	/omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:121 +0x29a
created by github.com/DataDog/datadog-agent/pkg/autodiscovery.NewAutoConfig
	/omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/autodiscovery/autoconfig.go:71 +0x65
```

Looking forward for the fix. Thanks for the quick reaction

**Comment by ykyr** [comment]
@hush-hush I added your suggestion to the dd config on one of the clusters, and since yesterday there are no restarts of the datadog-agents 
```
logs_config:
  cca_in_ad: true 
```

Not sure what exactly this configuration does, but looks like it helped. At the same time I'm looking forward to the new release with the fix. 
Thanks!

**Pull request merged**